pymc3/backends/arviz.py

```diff
@@ -0,0 +1,614 @@
+                self.attrs = {
+                    "sampling_time": trace.report.t_sampling,
+                    "tuning_steps": trace.report.n_tune,
+                }
+            else:
+                self.ndraws = len(trace)
+                if self.save_warmup:
+                    warnings.warn(
+                        "Warmup samples will be stored in posterior group and will not be"
+                        " excluded from stats and diagnostics."
+                        " Do not slice the trace manually before conversion",
+                        UserWarning,
+                    )
+            self.ntune = len(self.trace) - self.ndraws
+            self.posterior_trace, self.warmup_trace = self.split_trace()
+        else:
+            self.nchains = self.ndraws = 0
+
+        self.prior = prior
+        self.posterior_predictive = posterior_predictive
+        self.log_likelihood = log_likelihood
+        self.predictions = predictions
+        self.index_origin = rcParams["data.index_origin"] if index_origin is None else index_origin
+
+        def arbitrary_element(dct: Dict[Any, np.ndarray]) -> np.ndarray:
+            return next(iter(dct.values()))
+
+        if trace is None:
+            # if you have a posterior_predictive built with keep_dims,
+            # you'll lose here, but there's nothing I can do about that.
+            self.nchains = 1
+            get_from = None
+            if predictions is not None:
+                get_from = predictions
+            elif posterior_predictive is not None:
+                get_from = posterior_predictive
+            elif prior is not None:
+                get_from = prior
+            if get_from is None:
+                # pylint: disable=line-too-long
+                raise ValueError(
+                    "When constructing InferenceData must have at least"
+                    " one of trace, prior, posterior_predictive or predictions."
+                )
+
+            aelem = arbitrary_element(get_from)
+            self.ndraws = aelem.shape[0]
+
+        self.coords = {} if coords is None else coords
+        if hasattr(self.model, "coords"):
+            self.coords = {**self.model.coords, **self.coords}
+
+        self.dims = {} if dims is None else dims
+        if hasattr(self.model, "RV_dims"):
+            model_dims = {k: list(v) for k, v in self.model.RV_dims.items()}
+            self.dims = {**model_dims, **self.dims}
+
+        self.observations = self.find_observations()
+
+    def find_observations(self) -> Optional[Dict[str, Var]]:
+        """If there are observations available, return them as a dictionary."""
+        if self.model is None:
+            return None
+        observations = {}
+        for obs in self.model.observed_RVs:
+            if hasattr(obs.tag, "observations"):
+                aux_obs = obs.tag.observations
+                observations[obs.name] = aux_obs.data if hasattr(aux_obs, "data") else aux_obs
```

Comment:
Is `data` the right field? Or should it be `value`? Or try both?

Reply:
That will work for `Constant` objects, but not for shared variables, which will require `aux_obs.get_value()`.

There's also the case where `obs.tag.observations` is a partially observed variable (i.e. has missing data).  The result will be an `AdvancedIncSubtensor1` in that case.

Here's a helper function for all those situations:
```python
import numpy as np

from aesara.graph.basic import Constant
from aesara.tensor.sharedvar import SharedVariable
from aesara.tensor.subtensor import AdvancedIncSubtensor1


def extract_data(x):
    if isinstance(x, Constant):
        return x.data
    if isinstance(x, SharedVariable):
        return x.get_value()
    if x.owner and isinstance(x.owner.op, AdvancedIncSubtensor1):
        array_data = extract_data(x.owner.inputs[0])
        mask_idx = extract_data(x.owner.inputs[2])
        mask = np.zeros_like(array_data)
        mask[mask_idx] = 1
        return np.ma.MaskedArray(array_data, mask)

    raise TypeError(f"Data cannot be extracted from {x}")


```